app/models/user_batch.rb

```diff
@@ -35,79 +36,247 @@ def create_users(mission)
+      end.flatten]
+
+      user_batch_attributes << attributes
+    end
+
+    user_batch_attributes
+  end
+
+  def create_users_instances(attribs, mission)
+    attribs.map{ |attrib| create_new_user(attrib, mission) }
+  end
+
+  def create_hash_table_with_fields_and_indexes(objects)
+    @fields_hash_table = {
+      'email' => {},
+      'login' => {},
+      'phone' => {}
+    }
+
+    columns = ['email', 'login', 'phone', 'phone2']
+
+    columns.each{ |field| populate_hash_with_field_and_occurrences(field, objects) }
+  end
+
+  def populate_hash_with_field_and_occurrences(field, objects)
+    key = correct_field_key(field)
+
+    objects.map{|o| o.send(field)}.each_with_index do |value, index|
+      (@fields_hash_table[key][value] ||= []) << index unless value.nil?
+    end
+  end
+
+  def validate_users_batch(users_batch)
+    users_batch.each{|u| u.valid? }
+  end
+
+  def check_validation_errors(users, row_start)
+    users.each_with_index do |user, index|
+      row_number = actual_row_number(row_start, index)
+      add_validation_error_messages(user, row_number)
+
+      if errors_reached_limit
+        add_too_many_errors(row_number)
+        break
+      end
+    end
+  end
+
+  def create_new_user(attributes, mission)
+    User.new(attributes.merge(
+             reset_password_method: "print",
+             admin: false,
+             assignments: [Assignment.new(mission: mission, role: User::ROLES.first)],
+             skip_validations_with_queries: true,
+             import_num: @import_num += 1))
+  end
 
-        # skip blank rows
-        next if row.all?(&:blank?)
-
-        # turn the row into an attribute hash
-        attributes = Hash[*row.map do |cell|
-          field = fields[cell.coordinate.column - 1]
-          [field, cell.value.presence]
-        end.flatten]
-
-        # attempt to create the user with the parsed params
-        user = User.create(attributes.merge(
-          reset_password_method: "print",
-          assignments: [Assignment.new(mission_id: mission.id, role: User::ROLES.first)]))
-
-        # if the user has errors, add them to the batch's errors
-        if !user.valid?
-          user.errors.keys.each do |attribute|
-            user.errors.full_messages_for(attribute).each do |error|
-              row_error = I18n.t('import.row_error', row: row_number, error: error)
-              errors.add("users[#{row_index}].#{attribute}", row_error)
-            end
+  def add_validation_error_messages(user, row_number)
+    # if the user has errors, add them to the batch's errors
+    unless user.errors.empty?
+      user.errors.keys.each do |attribute|
+        if attribute != :persistence_token
+          user.errors.full_messages_for(attribute).each do |error|
+            add_error(error, attribute, row_number)
           end
-          @validation_error = true
         end
+      end
+      @validation_error = true unless error_is_on_persistence_token? user
```

Comment:
If you move this inside the `if` statement above you shouldn't need the extra method, right?